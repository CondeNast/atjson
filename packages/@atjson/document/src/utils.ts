import { Annotation } from "./internals";

/**
 * Checks whether an annotation is an instance
 * of the Annotation class.
 */
export function is<
  T extends {
    vendorPrefix: string;
    type: string;
    new (...args: any[]): Annotation<any>;
  },
>(annotation: Annotation<any>, Class: T): annotation is InstanceType<T> {
  return (
    annotation &&
    annotation.type === Class.type &&
    annotation.vendorPrefix === Class.vendorPrefix
  );
}
