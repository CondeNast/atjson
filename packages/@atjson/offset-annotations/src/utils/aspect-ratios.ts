const ASPECT_RATIOS = [
  // Full portrait
  "9:16",
  // Vertical
  "2:3",
  "3:4",
  "4:5",
  // Square
  "1:1",
  // Fox Movietone
  "6:5",
  // Early TV
  "5:4",
  // Traditional TV
  "4:3",
  // Academy standard film
  "11:8",
  // 35mm
  "3:2",
  // Used for 4:3 / 16:9 TV compatability
  "14:9",
  // Computer screens
  "16:10",
  // European widescreen standard / Paramount format / Super 16mm
  "5:3",
  // HD video standard / Digital broadcast TV standard
  "16:9",
  // US widescreen cinema standard
  "37:20",
  // 4K & 2K standard
  "19:10",
  // Univisium
  "2:1",
  // 70mm film
  "11:5",
  // Cinematic widescreen
  "21:9",
  // Widescreen cinema standard
  "12:5",
  // Ultra-WideScreen
  "18:5",
] as const;

const proportions = ASPECT_RATIOS.map((ratio) => {
  let [width, height] = ratio.split(":");
  return {
    ratio,
    proportion: parseFloat(width) / parseFloat(height),
  };
}).sort((a, b) => {
  return a.proportion < b.proportion ? -1 : a.proportion > b.proportion ? 1 : 0;
});

const horizontalAspectRatios = proportions.filter(
  ({ proportion }) => proportion >= 1
);
const verticalAspectRatios = proportions.filter(
  ({ proportion }) => proportion <= 1
);

/**
 * Returns the closest aspect ratio for a video,
 * given a width and height. This is used to get the aspect
 * ratio for the VideoEmbed
 */
export function getClosestAspectRatio(width: number, height: number) {
  let proportion = width / height;
  let delta = Infinity;

  if (proportion === 1) {
    return "1:1" as const;
  }

  let aspectRatios =
    proportion > 1 ? horizontalAspectRatios : verticalAspectRatios;

  for (let i = 0, len = aspectRatios.length; i < len; i++) {
    let aspectRatio = aspectRatios[i];
    let currentDelta = Math.abs(proportion - aspectRatio.proportion);

    if (currentDelta > delta) {
      return aspectRatios[i - 1].ratio;
    }
    delta = currentDelta;
  }
  return aspectRatios[aspectRatios.length - 1].ratio;
}
