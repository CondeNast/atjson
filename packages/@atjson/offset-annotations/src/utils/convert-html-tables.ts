import {
  AdjacentBoundaryBehaviour,
  JSON,
  SliceAnnotation,
  compareAnnotations,
} from "@atjson/document";
import OffsetSource, { DataSet, Table, ColumnType } from "../index";

export function convertHTMLTablesToDataSet(
  doc: OffsetSource,
  vendor: string
): void {
  doc.where({ type: `-${vendor}-table` }).forEach((table) => {
    let dataColumnHeaders: {
      name: string;
      slice: string;
      type: ColumnType;
    }[] = [];
    let columnConfigs: Exclude<Table["attributes"]["columns"], undefined> = [];
    let dataRows: Record<string, { slice: string; jsonValue: JSON }>[] = [];

    let slices: SliceAnnotation[] = [];

    doc
      .where(
        (annotation) =>
          annotation.vendorPrefix === vendor &&
          annotation.type === "th" &&
          annotation.start >= table.start &&
          annotation.end <= table.end
      )
      .sort(compareAnnotations)
      .forEach((headCell, index) => {
        if (index !== 0) {
          doc.insertText(
            headCell.start,
            " ",
            AdjacentBoundaryBehaviour.preserveBoth
          );
        }

        let slice = new SliceAnnotation({
          ...headCell,
          id: undefined,
          attributes: { refs: [] },
        });
        doc.replaceAnnotation(headCell, slice);
        let columnName = doc.content
          .slice(slice.start, slice.end)
          .replace(/[^\p{Letter}\p{White_Space}]/gu, "")
          .replace(/\s+/, " ");
        dataColumnHeaders.push({
          slice: slice.id,
          name: columnName.length ? columnName : `column ${index + 1}`,
          type: ColumnType.PERITEXT,
        });
        slices.push(slice);

        if (headCell.attributes.style) {
          let { groups }: RegExpMatchArray = headCell.attributes.style.match(
            /(text-align: ?(?<alignment>left|right|center))/
          );

          if (groups?.alignment) {
            columnConfigs.push({
              name: columnName,
              textAlign: groups?.alignment as "left" | "right" | "center",
            });
          } else {
            columnConfigs.push({ name: columnName });
          }
        } else {
          columnConfigs.push({ name: columnName });
        }
      });

    let tableRows = doc.where(
      (annotation) =>
        annotation.vendorPrefix === vendor &&
        annotation.type === "tr" &&
        annotation.start >= table.start &&
        annotation.end <= table.end
    );

    tableRows.forEach((row) => {
      doc.insertText(row.start, " ", AdjacentBoundaryBehaviour.preserveBoth);

      let rowEntries: [string, { slice: string; jsonValue: JSON }][] = [];
      doc
        .where(
          (annotation) =>
            annotation.vendorPrefix === vendor &&
            annotation.type === "td" &&
            annotation.start >= row.start &&
            annotation.end <= row.end
        )
        .sort(compareAnnotations)
        .forEach((bodyCell, index) => {
          if (index !== 0) {
            doc.insertText(
              bodyCell.start,
              " ",
              AdjacentBoundaryBehaviour.preserveBoth
            );
          }
          let slice = new SliceAnnotation({
            ...bodyCell,
            id: undefined,
            attributes: { refs: [] },
          });
          doc.replaceAnnotation(bodyCell, slice);
          rowEntries.push([
            dataColumnHeaders[index].name,
            {
              slice: slice.id,
              jsonValue: doc.content.slice(slice.start, slice.end),
            },
          ]);
          slices.push(slice);
        });

      if (rowEntries.length) {
        dataRows.push(Object.fromEntries(rowEntries));
      }
    });

    tableRows.remove();

    let dataSet = new DataSet({
      ...table,
      id: undefined,
      attributes: {
        columns: dataColumnHeaders,
        rows: dataRows,
      },
    });

    let offsetTable = new Table({
      ...table,
      id: undefined,
      attributes: { dataSet: dataSet.id },
    });

    if (columnConfigs.length) {
      offsetTable.attributes.columns = columnConfigs;
    }

    slices.forEach((slice) => slice.attributes.refs.push(dataSet.id));

    doc.replaceAnnotation(table, dataSet, offsetTable);
  });

  doc.where({ type: `-${vendor}-thead` }).remove();
  doc.where({ type: `-${vendor}-tbody` }).remove();
}
