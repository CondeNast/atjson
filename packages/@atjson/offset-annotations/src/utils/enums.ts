export enum TextAlignment {
  Start = "start",
  End = "end",
  Center = "center",
  Justify = "justify",
}
