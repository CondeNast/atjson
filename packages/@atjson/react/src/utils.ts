import { Block, Mark, InternalMark } from "./types";
import { BLOCK_MARKER, ROOT } from "./const";

export function extractSlices(value: {
  marks: Mark[];
  blocks: Block[];
  text: string;
}) {
  let marksWithoutSlices: InternalMark[] = [];
  let slices: InternalMark[] = [];
  for (let i = 0, len = value.marks.length; i < len; i++) {
    let mark = value.marks[i];
    let match = mark.range.match(/([[|(])(\d+)\.\.(\d+)([\]|)])/);
    if (match == null) {
      throw new Error(`Malformed range ${mark.range}`);
    }
    if (mark.type === "slice") {
      slices.push({
        start: parseInt(match[2]),
        end: parseInt(match[3]),
        ...mark,
      });
    } else {
      marksWithoutSlices.push({
        start: parseInt(match[2]),
        end: parseInt(match[3]),
        ...mark,
      });
    }
  }

  let sliceMap: Record<
    string,
    { text: string; marks: InternalMark[]; blocks: Block[] }
  > = {};

  let blockBoundaryPositions: number[] = [];
  let blockIndex = value.text.indexOf(BLOCK_MARKER);
  while (blockIndex !== -1) {
    blockBoundaryPositions.push(blockIndex);
    blockIndex = value.text.indexOf(BLOCK_MARKER, blockIndex + 1);
  }
  blockBoundaryPositions.push(value.text.length);

  // A list of text ranges to remove for the remainder
  // document. The remainder will be constructed after the
  // slices are extracted so we don't have duplicated text.
  let rangesToDelete: [number, number][] = [];

  for (let i = 0, len = slices.length; i < len; i++) {
    let slice = slices[i];
    let { start, end } = slice;

    let text = value.text.slice(start, end);

    let blocks: Block[] = [];
    for (let j = 0, jlen = blockBoundaryPositions.length; j < jlen; j++) {
      let position = blockBoundaryPositions[j];
      if (position < start) {
        // Keep searching forward
        continue;
      } else if (position + 1 <= end) {
        // Add the block to the slice blocks list
        let block = value.blocks[j];
        let parentIndex = (blocks[0] ?? block).parents.length;
        blocks.push({
          ...block,
          id: `${slice.id}-${block.id}`,
          parents: block.parents.slice(parentIndex),
        });
      } else {
        // If the block index is after the slice, we can safely break
        // from the loop, saving some extra work.
        break;
      }
    }

    let marks: InternalMark[] = [];
    for (let j = 0, jlen = marksWithoutSlices.length; j < jlen; j++) {
      let mark = marksWithoutSlices[j];

      // For slicing purposes, we only consider marks whose
      // boundaries are within the slice to be represented.
      if (
        mark.start >= start &&
        mark.start <= end &&
        mark.end >= start &&
        mark.end <= end
      ) {
        marks.push({
          ...mark,
          id: `${slice.id}-${mark.id}`,
          start: Math.max(mark.start - start, 0),
          end: Math.min(mark.end - end, end - start),
        });
      }
    }

    // After collecting blocks, we'll need to do a pass
    // over the text and blocks to ensure that it's well
    // formed and there's a block parent for all text.
    if (blocks.length === 0) {
      text = `${BLOCK_MARKER}${text}`;
      blocks.push({
        id: `text-${slice.id}`,
        type: "text",
        parents: [],
        selfClosing: false,
        attributes: {},
      });
    }

    sliceMap[slice.id] = {
      text,
      marks,
      blocks,
    };

    // After handling the slice map, we'll be handling the
    // remainder of the document that excludes slices.
    // For cases where the underlying range is retained,
    // the content will be kept.
    if (slice.attributes.retain) {
      // If the slice is retained, we get to skip this
      // step since we want marks to be kept intact
      continue;
    }

    let isRangeInserted = false;
    // Add the slice to the ranges to delete
    for (let j = 0, jlen = rangesToDelete.length; j < jlen; j++) {
      let rangeToDelete = rangesToDelete[j];

      if (start >= rangeToDelete[0] && start <= rangeToDelete[1]) {
        // Extend the end of the range if the start is
        // within the current range
        rangeToDelete[1] = Math.max(rangeToDelete[1], end);
        isRangeInserted = true;
      } else if (end >= rangeToDelete[0] && end <= rangeToDelete[1]) {
        // Extend the start of the range if the end is
        // within the current range
        rangeToDelete[0] = Math.min(rangeToDelete[0], start);
        isRangeInserted = true;
      }
    }
    // If the slice wasn't within a range or extended a range,
    // append it to the list of ranges to remove
    if (!isRangeInserted) {
      rangesToDelete.push([start, end]);
    }
  }

  let text = value.text.slice(0, rangesToDelete[0][0]);
  let lastEnd;
  for (let i = 0; i < rangesToDelete.length - 1; i++) {
    text += value.text.slice(rangesToDelete[i][1], rangesToDelete[i + 1][0]);
    lastEnd = rangesToDelete[i + 1][1];
  }

  text = text + value.text.slice(lastEnd);

  let marks: InternalMark[] = [];
  for (let mark of marksWithoutSlices) {
    let isKept = true;
    let startOffset = 0;
    let endOffset = 0;
    for (let range of rangesToDelete) {
      // If the mark is fully inside the range,
      // it won't be represented in the remainder document
      if (mark.start >= range[0] && mark.end <= range[1]) {
        isKept = false;
        break;
      }

      // Get range offset information
      // for marks that are added

      if (mark.start >= range[0] && mark.start <= range[1]) {
        // If the start is inside of the range,
        // clamp it to the range start.
        startOffset += mark.start - range[0];
      } else if (mark.start > range[1]) {
        // If the start is after the range,
        // we can add the whole range to the offset
        startOffset += range[1] - range[0];
      }

      if (mark.end >= range[0] && mark.end <= range[1]) {
        // If the end is inside of the range,
        // clamp it to the range start.
        endOffset += mark.end - range[0];
      } else if (mark.end > range[1]) {
        // If the end is after the range,
        // we can add the whole range to the offset
        endOffset += range[1] - range[0];
      }
    }

    if (isKept) {
      let start = mark.start - startOffset;
      let end = mark.end - endOffset;
      marks.push({
        ...mark,
        start,
        end,
        range: `${mark.range[0]}${start}..${end}${
          mark.range[mark.range.length - 1]
        }`,
      });
    }
  }

  let blocks: Block[] = [];
  for (let i = 0, len = blockBoundaryPositions.length; i < len; i++) {
    let position = blockBoundaryPositions[i];
    let isKept = true;
    for (let range of rangesToDelete) {
      isKept = isKept && !(position >= range[0] && position + 1 <= range[1]);
    }
    if (isKept && value.blocks[i]) {
      blocks.push(value.blocks[i]);
    }
  }

  return [
    {
      text,
      marks,
      blocks,
    },
    sliceMap,
  ] as const;
}

export function createTree(value?: {
  marks: InternalMark[];
  blocks: Block[];
  text: string;
}) {
  let chunks = (value?.text ?? "").split(BLOCK_MARKER);
  let blocks = value?.blocks ?? [];
  let marks = value?.marks ?? [];

  let tree: Record<string, Array<Block | InternalMark | string>> = {};

  let stack: Block[] = [];
  let blockStart = 1;
  for (let i = 0, len = blocks.length; i < len; i++) {
    let block = blocks[i];
    let text = chunks[i + 1];
    let blockEnd = blockStart + text.length;

    while (block.parents.length < stack.length) {
      stack.pop();
    }

    let scopedMarks: InternalMark[] = [];
    for (let j = 0, jlen = marks.length; j < jlen; j++) {
      let mark = marks[j];
      if (mark.start >= blockStart && mark.start <= blockEnd) {
        scopedMarks.push(mark);
      }
    }

    let parentId = stack[stack.length - 1]?.id ?? ROOT;
    let blockId =
      block.selfClosing || block.type === "text" ? parentId : block.id;

    tree[parentId] = tree[parentId] ?? [];
    if (block.type !== "text") {
      tree[parentId].push(block);
    }

    if (tree[blockId] == null) {
      tree[blockId] = [];
    }

    // If there's no marks, we can return the text as a single chunk
    if (scopedMarks.length === 0) {
      if (text.length) {
        tree[blockId].push(text);
      }
    } else {
      // To chunk the text properly, we need to handle several cases:
      // 1. Overlapping marks
      // 2. Nested marks
      // 3. Marks going outside the block boundary

      // First, we'll mark indexes where mark boundaries are
      let indexes: number[] = [];
      for (let i = 0, len = scopedMarks.length; i < len; i++) {
        let { start, end } = scopedMarks[i];
        if (indexes.indexOf(start) === -1 && start >= blockStart) {
          indexes.push(start);
        }
        if (indexes.indexOf(end) === -1 && end <= blockEnd) {
          indexes.push(end);
        }
      }
      indexes.sort((a, b) => a - b);

      // Now with the mark boundaries, we can construct a tree
      // of items in the block
      let start = blockStart;

      // We'll walk through the ranges and insert nodes into
      // the correct parent
      let markStack: InternalMark[] = [];
      for (let i = 0, len = indexes.length; i < len; i++) {
        let end = indexes[i];
        if (start == end) continue;
        // Then add marks that are starting from this position
        scopedMarks
          .filter((mark) => mark.start === start)
          .sort((a, b) => b.end - a.end)
          .forEach((mark) => {
            markStack.push(mark);
          });

        let parentId: string = blockId;
        for (let i = 0, len = markStack.length; i < len; i++) {
          let mark = markStack[i];
          let id = parentId === blockId ? mark.id : `${parentId}-${mark.id}`;

          // Check to see if the mark has already been inserted
          // into the tree. There's no cases where we'll have
          // marks that are disjoint at the same level of the
          // tree, since marks are all treated the same.
          let parent = tree[parentId];
          let isInserted = false;
          for (let j = 0, jlen = parent.length; j < jlen; j++) {
            let node = parent[j];
            isInserted =
              isInserted || (typeof node !== "string" && node.id === id);
          }

          if (!isInserted) {
            tree[parentId].push({
              ...mark,
              id,
            });
          }
          parentId = id;
          if (tree[parentId] == null) {
            tree[parentId] = [];
          }
        }

        let leaf = text.slice(start - blockStart, end - blockStart);
        if (leaf.length) {
          tree[parentId].push(leaf);
        }

        // Remove from the stack marks that have ended first
        scopedMarks
          .filter((mark) => mark.end === end)
          .forEach((mark) => {
            markStack.splice(markStack.indexOf(mark), 1);
          });
        start = end;
      }
      let remainder = text.slice(start - blockStart);
      if (remainder) {
        tree[blockId].push(remainder);
      }
    }

    stack.push(block);
    blockStart = blockEnd + 1;
  }
  return tree;
}
