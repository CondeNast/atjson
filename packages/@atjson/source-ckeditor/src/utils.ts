import * as CK from "./ckeditor";

export function isTextNode(
  node: CK.Node | CK.DocumentFragment,
): node is CK.TextNode {
  return node.is("text");
}

export function isRootElement(
  node: CK.Node | CK.DocumentFragment,
): node is CK.RootElement {
  return node.is("rootElement");
}

export function isElement(
  node: CK.Node | CK.DocumentFragment,
): node is CK.Element {
  return node.is("element");
}

export function isNode(node: CK.Node | CK.DocumentFragment): node is CK.Node {
  return node.is("node");
}

export function isDocumentFragment(
  node: CK.Node | CK.DocumentFragment,
): node is CK.DocumentFragment {
  return node.is("documentFragment");
}
