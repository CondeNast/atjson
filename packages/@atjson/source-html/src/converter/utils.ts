import { TextAlignment } from "@atjson/offset-annotations";

export function parseCSS(css: string | undefined) {
  let rules = (css || "").split(";");
  let styles: Record<string, string> = {};
  for (let rule of rules) {
    let [name, ...value] = rule.split(":");
    if (value.length > 0) {
      styles[name] = value.join(":").trim();
    }
  }
  return styles;
}

const RightToLeftAlignment = {
  start: TextAlignment.Start,
  left: TextAlignment.End,
  right: TextAlignment.Start,
  end: TextAlignment.End,
  center: TextAlignment.Center,
  justify: TextAlignment.Justify,
} as { [key: string]: TextAlignment | undefined };

const LeftToRightAlignment = {
  start: TextAlignment.Start,
  left: TextAlignment.Start,
  right: TextAlignment.End,
  end: TextAlignment.End,
  center: TextAlignment.Center,
  justify: TextAlignment.Justify,
} as { [key: string]: TextAlignment | undefined };

/**
 *
 * @param alignment The `text-align` property of the
 * @param language The language code of
 */
export function toTextAlignment(
  alignment: string | undefined,
  direction: string | undefined
) {
  if (alignment == null) {
    return alignment;
  }

  if (direction === "rtl") {
    return RightToLeftAlignment[alignment];
  }
  return LeftToRightAlignment[alignment];
}
