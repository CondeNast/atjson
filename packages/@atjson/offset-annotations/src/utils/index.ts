import { getClosestAspectRatio } from "./aspect-ratios";
import * as SocialURLs from "./social-urls";
import * as VideoURLs from "./video-urls";
import { convertHTMLTablesToDataSet } from "./convert-html-tables";

export {
  convertHTMLTablesToDataSet,
  getClosestAspectRatio,
  SocialURLs,
  VideoURLs,
};
