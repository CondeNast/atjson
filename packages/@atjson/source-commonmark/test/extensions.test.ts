import {
  InlineAnnotation,
  getConverterFor,
  is,
  serialize,
} from "@atjson/document";
import OffsetSource from "@atjson/offset-annotations";
import MarkdownIt from "markdown-it";
import CommonmarkSource from "../src";
import { render } from "./utils";

class StrikeThrough extends InlineAnnotation {
  static vendorPrefix = "commonmark";
  static type = "s";
}

class MarkdownItSource extends CommonmarkSource {
  static contentType = "application.vnd.atjson+markdownit";
  static schema = [...CommonmarkSource.schema, StrikeThrough];
  static get markdownParser() {
    return MarkdownIt();
  }
}

MarkdownItSource.defineConverterTo(OffsetSource, (doc) => {
  let convertCommonmark = getConverterFor(CommonmarkSource, OffsetSource);
  convertCommonmark(doc);
  doc.where({ type: "-commonmark-s" }).set({ type: "-offset-strikethrough" });

  return doc;
});

describe("strikethrough", () => {
  test("~~hello~~ is converted to strikethrough annotations", () => {
    let doc = MarkdownItSource.fromRaw("~~hello~~").withStableIds();
    expect(render(doc)).toBe("hello\n\n");
    let strikeThrough = doc.where((a) => is(a, StrikeThrough));
    expect(strikeThrough.toJSON()).toMatchInlineSnapshot(`
      [
        {
          "attributes": {},
          "end": 7,
          "id": "00000003",
          "start": 1,
          "type": "-commonmark-s",
        },
      ]
    `);
  });

  test("conversion to Offset uses existing conversions", () => {
    let doc = MarkdownItSource.fromRaw("~~hello~~ *world*")
      .convertTo(OffsetSource)
      .withStableIds();
    expect(
      doc
        .where((a) => a.type !== "parse-token")
        .sort()
        .toJSON()
    ).toMatchInlineSnapshot(`
      [
        {
          "attributes": {},
          "end": 14,
          "id": "00000001",
          "start": 0,
          "type": "-offset-paragraph",
        },
        {
          "attributes": {},
          "end": 7,
          "id": "00000003",
          "start": 1,
          "type": "-offset-strikethrough",
        },
        {
          "attributes": {},
          "end": 14,
          "id": "00000005",
          "start": 8,
          "type": "-offset-italic",
        },
      ]
    `);
  });
});

describe("tables", () => {
  const tableExample = `
| name     | age | job     | [*notes*](https://en.wikipedia.org/wiki/Delicious_in_Dungeon)                                                                                    |
|:-------- | ---:| ------- |:------------------------------------------------------------------------------------------------------------------------------------------------:|
| laios    | 20  | fighter | A strange but earnest person. He *really __really__* likes monsters                                                                              |
| marcille | 500 | mage    | Difficult to get along with but very competent. Despite seeming strict and fussy, she is interested in forbidden magic...                        |
| falin    | 18  | healer  | She *seems* nice, but is actually just a people pleaser. When push comes to shove she will look out for people she loves and disregard strangers |
| chilchuk | 29  | thief   | Looks like a child but is actually a divorced father of three. He is serious about his work and isn't interested in getting close with people    |
`;
  test("parsing", () => {
    expect(
      serialize(MarkdownItSource.fromRaw(tableExample), { withStableIds: true })
    ).toMatchSnapshot();
  });

  test("converting", () => {
    let doc = MarkdownItSource.fromRaw(tableExample).withStableIds();

    expect(
      serialize(doc.convertTo(OffsetSource), { withStableIds: true })
    ).toMatchSnapshot();
  });
});
