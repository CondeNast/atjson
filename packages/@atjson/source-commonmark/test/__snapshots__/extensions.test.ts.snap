// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tables converting 1`] = `
{
  "blocks": [
    {
      "attributes": {
        "columns": [
          {
            "name": "￼name",
            "slice": "M00000001",
            "type": "peritext",
          },
          {
            "name": "￼age",
            "slice": "M00000002",
            "type": "peritext",
          },
        ],
        "rows": [
          {
            "￼age": {
              "jsonValue": "￼20",
              "slice": "M00000004",
            },
            "￼name": {
              "jsonValue": "￼laios",
              "slice": "M00000003",
            },
          },
          {
            "￼age": {
              "jsonValue": "￼500",
              "slice": "M00000006",
            },
            "￼name": {
              "jsonValue": "￼marcille",
              "slice": "M00000005",
            },
          },
        ],
      },
      "id": "B00000000",
      "parents": [],
      "selfClosing": false,
      "type": "data-set",
    },
    {
      "attributes": {
        "dataSet": "M00000000",
      },
      "id": "B00000001",
      "parents": [
        "data-set",
      ],
      "selfClosing": false,
      "type": "table",
    },
  ],
  "marks": [
    {
      "attributes": {
        "refs": [
          "B00000001",
        ],
      },
      "id": "M00000000",
      "range": "(0..33]",
      "type": "slice",
    },
    {
      "attributes": {
        "refs": [
          "B00000000",
        ],
      },
      "id": "M00000001",
      "range": "(3..7]",
      "type": "slice",
    },
    {
      "attributes": {
        "refs": [
          "B00000000",
        ],
      },
      "id": "M00000002",
      "range": "(8..11]",
      "type": "slice",
    },
    {
      "attributes": {
        "refs": [
          "B00000000",
        ],
      },
      "id": "M00000003",
      "range": "(12..17]",
      "type": "slice",
    },
    {
      "attributes": {
        "refs": [
          "B00000000",
        ],
      },
      "id": "M00000004",
      "range": "(18..20]",
      "type": "slice",
    },
    {
      "attributes": {
        "refs": [
          "B00000000",
        ],
      },
      "id": "M00000005",
      "range": "(21..29]",
      "type": "slice",
    },
    {
      "attributes": {
        "refs": [
          "B00000000",
        ],
      },
      "id": "M00000006",
      "range": "(30..33]",
      "type": "slice",
    },
  ],
  "text": "￼￼ name age laios 20 marcille 500",
}
`;

exports[`tables parsing 1`] = `
{
  "blocks": [
    {
      "attributes": {},
      "id": "B00000000",
      "parents": [],
      "selfClosing": false,
      "type": "table",
    },
    {
      "attributes": {},
      "id": "B00000001",
      "parents": [
        "table",
      ],
      "selfClosing": false,
      "type": "thead",
    },
    {
      "attributes": {},
      "id": "B00000002",
      "parents": [
        "table",
        "thead",
      ],
      "selfClosing": false,
      "type": "tr",
    },
    {
      "attributes": {},
      "id": "B00000003",
      "parents": [
        "table",
        "thead",
        "tr",
      ],
      "selfClosing": false,
      "type": "th",
    },
    {
      "attributes": {},
      "id": "B00000004",
      "parents": [
        "table",
        "thead",
        "tr",
      ],
      "selfClosing": false,
      "type": "th",
    },
    {
      "attributes": {},
      "id": "B00000005",
      "parents": [
        "table",
      ],
      "selfClosing": false,
      "type": "tbody",
    },
    {
      "attributes": {},
      "id": "B00000006",
      "parents": [
        "table",
        "tbody",
      ],
      "selfClosing": false,
      "type": "tr",
    },
    {
      "attributes": {},
      "id": "B00000007",
      "parents": [
        "table",
        "tbody",
        "tr",
      ],
      "selfClosing": false,
      "type": "td",
    },
    {
      "attributes": {},
      "id": "B00000008",
      "parents": [
        "table",
        "tbody",
        "tr",
      ],
      "selfClosing": false,
      "type": "td",
    },
    {
      "attributes": {},
      "id": "B00000009",
      "parents": [
        "table",
        "tbody",
      ],
      "selfClosing": false,
      "type": "tr",
    },
    {
      "attributes": {},
      "id": "B0000000a",
      "parents": [
        "table",
        "tbody",
        "tr",
      ],
      "selfClosing": false,
      "type": "td",
    },
    {
      "attributes": {},
      "id": "B0000000b",
      "parents": [
        "table",
        "tbody",
        "tr",
      ],
      "selfClosing": false,
      "type": "td",
    },
  ],
  "marks": [],
  "text": "￼￼￼￼name￼age￼￼￼laios￼20￼￼marcille￼500",
}
`;
